File: CommentBanner/app/src/main/java/com/kuanquan/commentbanner/v2/IndicatorDrawView.java
//package com.kuanquan.commentbanner.v2;
//
//import android.content.Context;
//import android.graphics.drawable.Drawable;
//import android.util.AttributeSet;
//import android.view.Gravity;
//import android.view.View;
//import android.view.ViewGroup;
//import android.widget.FrameLayout;
//import android.widget.RelativeLayout;
//
//import androidx.appcompat.widget.AppCompatImageView;
//import androidx.appcompat.widget.LinearLayoutCompat;
//
//import com.blankj.utilcode.util.SizeUtils;
//
//public class IndicatorDrawView extends LinearLayoutCompat implements LIndicator {
//    private Context context;
//
//    private Drawable selectColor;
//    private Drawable normalColor;
//    private float alpha;
//
//    public IndicatorDrawView(Context context) {
//        super(context);
//        this.context = context;
//    }
//
//    public IndicatorDrawView(Context context, AttributeSet attrs) {
//        super(context, attrs);
//        this.context = context;
//    }
//
//    public IndicatorDrawView(Context context, AttributeSet attrs, int defStyleAttr) {
//        super(context, attrs, defStyleAttr);
//        this.context = context;
//    }
//
//    public IndicatorDrawView initView(int indicatorCount,int selectPosition,Drawable selectDrawble,Drawable normalDrawble,float alpha){
//        this.selectColor = selectDrawble;
//        this.normalColor = normalDrawble;
//        this.alpha = alpha;
//        for (int i=0;i<indicatorCount;i++){
//            AppCompatImageView ivIndicator = new AppCompatImageView(context);
//            LinearLayoutCompat.LayoutParams lp = new LinearLayoutCompat.LayoutParams(LayoutParams.WRAP_CONTENT,LayoutParams.WRAP_CONTENT);
//            lp.leftMargin = i == 0 ? 0 : SizeUtils.dp2px(6);
//            ivIndicator.setLayoutParams(lp);
//            ivIndicator.setBackgroundDrawable(i == selectPosition ? selectDrawble : normalDrawble);
//            ivIndicator.setAlpha(i == selectPosition ? 1 : alpha);
//            addView(ivIndicator);
//        }
//        return this;
//    }
//
//    // 在Viewpage的监听中改变选择的指示器 改变指示器选中项
//    public void changeIndicator(int position){
//        int count = getChildCount();
//        for (int i=0;i<count;i++){
//            AppCompatImageView ivIndecator  = (AppCompatImageView) getChildAt(i);
//            ivIndecator.setBackgroundDrawable(normalColor);
//            ivIndecator.setAlpha(alpha);
//        }
//        AppCompatImageView ivIndecator  = (AppCompatImageView) getChildAt(position);
//        ivIndecator.setBackgroundDrawable(selectColor);
//        ivIndecator.setAlpha(1.0f);
//    }
//
//    @Override
//    public void initIndicatorCount(int pagerCount) {
//        setVisibility(pagerCount > 1 ? VISIBLE : GONE);
//        requestLayout();
//    }
//
//    @Override
//    public View getView() {
//        return this;
//    }
//
//    /**
//     * 控制在banner中的位置
//     */
//    private RelativeLayout.LayoutParams params;
//    @Override
//    public RelativeLayout.LayoutParams getParams() {
//        if (params == null) {
//            params = new RelativeLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
//            params.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM);
//            params.addRule(RelativeLayout.CENTER_HORIZONTAL);
//            params.bottomMargin = dip2px(10);
//        }
//        return params;
//    }
//
//    /**
//     * 控制在banner中的位置
//     */
//    private FrameLayout.LayoutParams frameLayoutParams;
//    @Override
//    public FrameLayout.LayoutParams getFlParams() {
//        if (frameLayoutParams == null) {
//            frameLayoutParams = new FrameLayout.LayoutParams(FrameLayout.LayoutParams.WRAP_CONTENT, FrameLayout.LayoutParams.WRAP_CONTENT);
//            frameLayoutParams.gravity = Gravity.BOTTOM | Gravity.CENTER_HORIZONTAL;
//            frameLayoutParams.bottomMargin = dip2px(8);
//        }
//        return frameLayoutParams;
//    }
//
//    @Override
//    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
//
//    }
//
//    @Override
//    public void onPageSelected(int position) {
//        changeIndicator(position);
//    }
//
//    @Override
//    public void onPageScrollStateChanged(int state) {
//
//    }
//
//    private int dip2px(float dp) { return (int) (dp * getContext().getResources().getDisplayMetrics().density); }
//
//    public static class Builder{
//        private Context context;
//        private int position;
//        private int indicatorCount;
//        private Drawable selectColor;
//        private Drawable normalColor;
//        private float alpha;
//
//        public Builder setContext(Context context){
//            this.context = context;
//            return this;
//        }
//
//        public Builder setSelectPosition(int position){
//            this.position = position;
//            return this;
//        }
//
//        public Builder setIndicatorCount(int indicatorCount){
//            this.indicatorCount = indicatorCount;
//            return this;
//        }
//
//        public Builder setSelectColor(Drawable selectColor){
//            this.selectColor = selectColor;
//            return this;
//        }
//
//        public Builder setNormalColor(Drawable normalColor){
//            this.normalColor = normalColor;
//            return this;
//        }
//
//        public Builder setNormalAlpha(float alpha){
//            this.alpha = alpha;
//            return this;
//        }
//
//        public IndicatorDrawView builder(){
//            return  new IndicatorDrawView(context).initView(indicatorCount,position,selectColor,normalColor,alpha);
//        }
//    }
//}

File: CommentBanner/app/src/main/java/com/kuanquan/commentbanner/v2/IndicatorDrawView.kt
package com.kuanquan.commentbanner.v2

import android.content.Context
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.*
import android.widget.*
import androidx.appcompat.widget.*
import com.blankj.utilcode.util.SizeUtils

class IndicatorDrawView(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : LinearLayoutCompat(context, attrs, defStyleAttr), LIndicator {
    private var selectDrawable: Drawable? = null
    private var normalDrawable: Drawable? = null
    private var alphaDrawable = 0f

    /**
     * param selectPosition 初始选择第几个
     * param indicatorCount 指示器总数量
     */
    fun startIndicator(selectPosition: Int, indicatorCount: Int) {
        for (i in 0 until indicatorCount) {
            val ivIndicator = AppCompatImageView(context)
            val lp = LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT)
            lp.leftMargin = if (i == 0) 0 else SizeUtils.dp2px(6f)
            ivIndicator.layoutParams = lp
            ivIndicator.setBackgroundDrawable(if (i == selectPosition) selectDrawable else normalDrawable)
            ivIndicator.alpha = if (i == selectPosition) 1f else alphaDrawable
            addView(ivIndicator)
        }
    }

    // 在Viewpage的监听中改变选择的指示器 改变指示器选中项
    fun changeIndicator(position: Int) {
        val count = childCount
        for (i in 0 until count) {
            val ivIndecator = getChildAt(i) as AppCompatImageView
            ivIndecator.setBackgroundDrawable(normalDrawable)
            ivIndecator.alpha = alphaDrawable
        }
        val ivIndecator = getChildAt(position) as AppCompatImageView
        ivIndecator.setBackgroundDrawable(selectDrawable)
        ivIndecator.alpha = 1.0f
    }

    override fun initIndicatorCount(pagerCount: Int) {
        visibility = if (pagerCount > 1) View.VISIBLE else View.GONE
        requestLayout()
    }

    override fun getView(): View {
        return this
    }

    /**
     * 控制在banner中的位置
     */
    private var frameLayoutParams: FrameLayout.LayoutParams? = null
    override fun getParams(): FrameLayout.LayoutParams? {
        if (frameLayoutParams == null) {
            frameLayoutParams = FrameLayout.LayoutParams(FrameLayout.LayoutParams.WRAP_CONTENT, FrameLayout.LayoutParams.WRAP_CONTENT)
            frameLayoutParams?.gravity = Gravity.BOTTOM or Gravity.CENTER_HORIZONTAL
            frameLayoutParams?.bottomMargin = dip2px(8f)
        }
        return frameLayoutParams
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}
    override fun onPageSelected(position: Int) {
        changeIndicator(position)
    }

    override fun onPageScrollStateChanged(state: Int) {}
    private fun dip2px(dp: Float): Int {
        return (dp * context.resources.displayMetrics.density).toInt()
    }

    fun setSelectColor(selectDrawable: Drawable?) {
        this.selectDrawable = selectDrawable
    }

    fun setNormalColor(normalDrawable: Drawable?) {
        this.normalDrawable = normalDrawable
    }

    fun setNormalAlpha(alphaDrawable: Float) {
        this.alphaDrawable = alphaDrawable
    }
    
}